File: 0137 Single Number II/single_number_iii.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
public:
    void binCalc(vector<int>& powArr, int a) {
        for(int i=0; i<32; i++) {
            int curr = a & (1<<i);
            if(curr) powArr[i]++;
        }
    }
    int convBinDec(vector<int>& powArr) {
        int num=0;
        for(int i=0; i<32; i++) {
            num+= (powArr[i]*(1<<i));
        }
        return num;
    }
    int singleNumber(vector<int>& nums) {
        vector<int> powArr(32,  0);
        for(auto& a:nums) binCalc(powArr, a);
        for(int i=0; i<32; i++) powArr[i]%=3;
        return convBinDec(powArr);
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 260-single-number-iii/single-number-iii.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public int[] singleNumber(int[] nums) {
        int xor = 0;
        for(int num:nums) {
            xor ^= num;
        }
        int msk = 1;
        while( (xor&msk) == 0 ) {
            msk<<=1;
        }

        int n1 = 0;
        int n2 = 0;
        for(int num:nums) {
            if((num&msk) == 0) {
                n1 ^= num;
            } else {
                n2 ^= num;
            }
        }

        return new int[]{n1, n2};
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
